base/iterators.jl

```diff
@@ -1095,4 +1094,41 @@ eltype(::Type{Stateful{T, VS}} where VS) where {T} = eltype(T)
 IteratorEltype(::Type{Stateful{T,VS}}) where {T,VS} = IteratorEltype(T)
 length(s::Stateful) = length(s.itr) - s.taken
 
+"""
+    only(x)
+
+Returns the one and only element of collection `x`, and throws an `ArgumentError` if the
+collection has zero or multiple elements.
+
+See also: [`first`](@ref), [`last`](@ref).
+
+!!! compat "Julia 1.4"
+    This method requires at least Julia 1.4.
+"""
+@propagate_inbounds function only(x)
+    i = iterate(x)
+    @boundscheck if i === nothing
+        throw(ArgumentError("Collection is empty, must contain exactly 1 element"))
+    end
+    (ret, state) = i
+    @boundscheck if iterate(x, state) !== nothing
+        throw(ArgumentError("Collection has multiple elements, must contain exactly 1 element"))
+    end
+    return ret
+end
+
+# Collections of known size
+only(x::Ref) = x[]
+only(x::Number) = x
+only(x::Char) = x
+only(x::Tuple{Any}) = x[1]
+only(x::Tuple) = throw(
+    ArgumentError("Tuple contains $(length(x)) elements, must contain exactly 1 element")
+)
+only(a::AbstractArray{<:Any, 0}) = @inbounds return a[]
```

Comment:
btw, this can still throw an `UndefRefError` -- which match `first`, but just flagging up the behaviour to be explcit (I think it is fine/good)

```julia
julia> empty_int = Array{Int, 0}(undef)
0-dimensional Array{Int64,0}:
1

julia> empty_str = Array{String, 0}(undef)
0-dimensional Array{String,0}:
#undef

julia> only(empty_int)
1

julia> first(empty_int)
1

julia> only(empty_str)
ERROR: UndefRefError: access to undefined reference
Stacktrace:
 [1] getindex at ./array.jl:728 [inlined]
 [2] _getindex at ./abstractarray.jl:1004 [inlined]
 [3] getindex at ./abstractarray.jl:981 [inlined]
 [4] only(::Array{String,0}) at ./REPL[100]:1
 [5] top-level scope at REPL[131]:1

julia> first(empty_str)
ERROR: UndefRefError: access to undefined reference
Stacktrace:
 [1] getindex at ./array.jl:728 [inlined]
 [2] first(::Array{String,0}) at ./abstractarray.jl:324
 [3] top-level scope at REPL[132]:1

```

Reply:
`only(Ref{String}())` also throws an undefined reference error. I'm not sure which is best and have no strong opinion here. But `only` dosctring says specifically which exception (`ArgumentError`) is thrown when there is not exactly one element, so this is a bit unsatisfying.